src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Date.cs

```diff
@@ -217,28 +223,30 @@ private static bool TryParseDateTimeOffset(ReadOnlySpan<byte> source, out DateTi
 
             sourceIndex = 20;
 
+            // Parse fraction
             {
+                int numDigitsRead = 0;
+
                 while (sourceIndex < source.Length && IsDigit(curByte = source[sourceIndex]))
                 {
                     int prevFractionTimesTen = fraction * 10;
 
-                    if (!(prevFractionTimesTen + (int)(curByte - (uint)'0') <= JsonConstants.MaxDateTimeFraction))
+                    if ((prevFractionTimesTen + (int)(curByte - (uint)'0') <= JsonConstants.MaxDateTimeFraction) && (numDigitsRead < JsonConstants.DateTimeNumFractionDigits))
```

Comment:
I am curious, what test case(s) required this change of tracking `numDigitsRead`? I am asking because maybe this condition can be simplified, especially since its within a loop.

Reply:
Cases with the fractional part < 0.1 i.e leading zeros e.g. `...Thh:mm.ss.01`, `ss.0009` etc.

- If we don't track how many digits we've seen, the test for if the fraction is <= than the max, `9_999_999` will never fail which means we could get digits beyond 7dp e.g. with `ss.00000000000001`

- We need to know exactly how many times to multiply the fraction by to get 7dp. Previously a string with of `ss.00001` would yield a fraction of `1000000` because we are multiplying till it's >= `9_999_999` instead of knowing to stop at `100`